# Zadanie otwarte

Jesteś pracownikiem w firmie spedycyjnej "TEU Logistics". Twoja firma zajmuje się przewozem towarów w morskich kontenerach własnymi statkami po całym świecie. Manager poprosił Cię o małą analizę danych, dlatego wysłał Ci plik z danymi (CSV), w którym są dane aktualnie transportowanych kontenerów przez "TEU Logistics". Dane są trochę mało czytelne i jest ich za dużo na "ręczne" kalkulacje.

W pierwszym wierszu, w każdej kolumnie podany jest numer, nazwa i klasa statku, który przewozi ładunek. W kolejnych wierszach, w każdej kolumnie widnieją dane kontenerów znajdujących się na pokładzie danego statu w kolumnie wraz z kwotą jaka została pobrana za przewóz kontenera.

Szczegóły: Nazwa statku, jego port startowy oraz docelowy i klasa podana jest w formacie:

aa-bb: xxxxxx (ttttttt)
gdzie:

aa - kraj pochodzenia kontenera
bb - kraj docelowy kontenera
xxxxxx: nazwa statku
tttttt: klasa statku
np: ES-PL: Cool Ship (Small Feeder)

Każdy kontener ma identyfikator nadany w formacie:

aa-bb-cccccccc/yyyy/xx@ddddddddd.ee
gdzie:

aa - kraj pochodzenia kontenera
bb - kraj docelowy kontenera
ccccccccc - numer kontenera (nie jest unikalny!!)
yyyy - waga kontenera w kilogramach (cyfrowo 0001-9999)
xx - typ ładunku w kontenerze (A0-Z9)
dddddddd.ee - nazwa i kraj pochodzenia firmy, która nadaje kontener, kraj jest 2 literowym oznaczeniem wg ISO 3166-1
długość pól cccccc, dddddd nie jest stała
pl-jp-2343432/2201/A1@companyname.pl oraz pl-jp-1223123/2201/A1@companyname.de to kontenery 2 różnych firm wg systemu informatycznego Twojej firmy.
Rekord jest połączeniem identyfikatora kontenera i kwoty w formacie:

identyfikator/kwota:
pl-jp-2343432/3100/Z1@companyname.pl/83427
Z wszelakich powodów takich jak różne kontrakty, terminy, kolejność załadunku, waga i rodzaj ładunku, cena wysyłki kontenera jest bardzo różna. Zauważ, że kraj pochodzenia i destynacja kontenera nie jest tożsamy z portem docelowym dla kontenera. Często taki kontenery trzeba przeładować kilka razy.

CSV: Format zapisu danych w którym rekordy oddzielone są od siebie znakiem średnika ';' a wiersze znakiem nowej linii '\n'

Odpowiedziami na zadanie są liczby lub jedno konkretne słowo. Odpowiedzi zawierające kalkulacje, kod, słowne zapisy liczb czy komentarze nie będą brane pod uwagę i będą oceniane na 0 punktów. Zadania można rozwiązać w dowolny sposób, a znajomość języka Python nie jest wymagana. Zadanie można rozwiązać w Excel lub dowolnym innym arkuszu kalkulacyjnym, z wykorzystaniem własnego programu napisanego w dowolnym języku programowania lub skryptem Basha z wykorzystaniem bibliotek do czytania plików CSV.

## PYTANIE NR 1 Ile kontenerów finalnie trafi do Japonii (JP)? (liczba kontenerów) – 15 pkt

In [1]:
import pandas as pd
import re

df = pd.read_csv('dane.csv', sep=';')
df

D:\conda\lib\site-packages\IPython\core\interactiveshell.py:3071: DtypeWarning: Columns (0,1,3,4,5,7,8,9,11,12,13,14,15,16,17,18,19,20,22,23,24,25,26,27,28,29,32,33,34,35,37,38,41,42,44,45,46,47,48) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


LK-SG Nellie (Feedermax)  \
0        MA-SG-83996625/2585/A1@oxlyo.uk/6145   
1      LK-SG-65041225/2292/X2@kldkrh.hr/99361   
2       LK-SG-56103708/2888/Z1@husvb.cl/93519   
3        LK-SG-34905922/2908/A2@jeck.ar/67001   
4       LK-SG-70365073/2973/B3@uexkc.ne/32974   
...                                       ...   
23957                                     NaN   
23958                                     NaN   
23959                                     NaN   
23960                                     NaN   
23961                                     NaN   

                     PA-OM Ghost (New Panamax)  \
0          PA-ES-96031647/2993/B2@tuw.ml/94044   
1      PA-OM-51292344/2432/Y2@cnnlgrt.ml/11242   
2          SA-SA-73465451/1924/B3@rej.ba/38170   
3        PA-ID-94657093/2370/X2@jpllu.ls/93944   
4         PA-OM-34467231/4591/Y2@hoox.ca/18710   
...                                        ...   
23957                                      NaN   
23958                                      NaN   
23959                                      NaN   
23960                                      NaN   
23961                                      NaN   

                                GR-BE Blue (ULCV)  \
0       MY-BE-40283751/4079/X1@nxocoezoq.mp/96326   
1          GR-BE-20853894/2657/Y2@rlzkic.dk/19799   
2            SA-GB-27472314/1430/Y2@vwit.ne/62695   
3            GR-BE-26706350/3679/X1@tjpy.ua/92050   
4        GR-BE-66955426/2830/Y2@ssaysqgy.nz/81484   
...                                           ...   
23957        GR-BE-58342500/3852/A2@byec.gs/74591   
23958  GR-BE-50084993/1223/B1@ndfwzbijal.hr/73698   
23959    KP-BE-87437408/3352/Y1@zpkxbcmc.uk/21191   
23960     GR-BE-14494354/4048/B3@apnggqi.tf/71909   
23961        GR-BE-47024355/1641/A1@tokm.td/48270   

                ID-VN Charlotte (Small feeder)  \
0       ID-VN-33266337/3165/Y2@uineka.se/69047   
1        BE-BE-29634771/2841/Z1@aerrv.pa/98217   
2      ID-VN-85411865/2549/B2@upthpgw.sb/87360   
3        ID-VN-15537738/3270/Y2@dtpcc.my/64445   
4         VN-IN-06555743/2249/Y1@ltor.ne/40761   
...                                        ...   
23957                                      NaN   
23958                                      NaN   
23959                                      NaN   
23960                                      NaN   
23961                                      NaN   

                   BE-CA Belafonte (Post-Panamax)  \
0            BE-CA-89905813/3223/C2@ltor.ne/68920   
1            BE-CA-38311826/3576/B3@jbkk.bf/71926   
2          BE-CA-42006271/1499/C1@pvvvtv.my/92789   
3      BE-CA-02105662/4492/Y1@irmqicyeym.cl/91984   
4         TH-CA-75372193/1736/Z1@pzhdqej.mn/13050   
...                                           ...   
23957                                         NaN   
23958                                         NaN   
23959                                         NaN   
23960                                         NaN   
23961                                         NaN   

                  MT-TH Speranza (New Panamax)  \
0      MT-TH-24923449/3639/C1@gpkapat.ge/53411   
1      DE-TH-62431607/1371/A1@qclajsg.ba/65645   
2        PH-TH-23141480/4337/C2@fukcw.sh/76034   
3        IN-LK-92283518/2950/A2@bqsbs.nz/38914   
4          MT-TH-15078686/3388/A1@mju.co/56044   
...                                        ...   
23957                                      NaN   
23958                                      NaN   
23959                                      NaN   
23960                                      NaN   
23961                                      NaN   

                               GR-ES China (ULCV)  \
0      ES-ES-05199984/1845/B2@recfkuwwsb.ml/19278   
1            GR-ES-74799208/4373/A2@srbh.cl/51634   
2            VN-MT-71958819/4118/X2@ofjx.af/42364   
3           PA-ES-28310638/4323/C3@psrun.ca/96238   
4      SA-ES-13902353/2942/A3@hsxpkywomc.na/86488   
...                                  

In [2]:
total = 0
for col in df.columns:
    total += df[col].str.count(r'^[A-Za-z]+-JP').sum()
    
print(total)

3779.0


Do Japonii trafi 4216 kontenerów

## PYTANIE NR. 2 Jaka klasa statku średnio przewozi najwięcej kontenerów? (sama nazwa klasy statku) – 15 pkt

In [3]:
df.columns

Index(['LK-SG Nellie (Feedermax)', 'PA-OM Ghost (New Panamax)',
       'GR-BE Blue (ULCV)', 'ID-VN Charlotte (Small feeder)',
       'BE-CA Belafonte (Post-Panamax)', 'MT-TH Speranza (New Panamax)',
       'GR-ES China (ULCV)', 'VN-SA Corsair (New Panamax)',
       'ES-TW Altair (Feedermax)', 'SA-ES Tankadère (Feedermax)',
       'KP-MY Gotha (ULCV)', 'ID-BR Cassidy (Feedermax)',
       'GR-TH Nautilus (Post-Panamax)', 'MT-MY Harekaze (Feeder)',
       'VN-VN Neptune (Panamax)', 'DE-MY Argonaut (New Panamax)',
       'KP-MY Odessa (Small feeder)', 'JP-IN Anchises (Panamax)',
       'MY-DE Seagallop (Feeder)', 'MA-ES Grayfish (Post-Panamax)',
       'CA-KP Reaper (Feeder)', 'KP-NL Tasha (ULCV)',
       'GR-BE Aeolus (Post-Panamax)', 'AE-IN Illustria (Small feeder)',
       'BE-GR Argonautica (Feedermax)', 'PA-ES Pocahontas (Small feeder)',
       'KP-GR Covenant (Feedermax)', 'MA-PA Ning-Po (Feeder)',
       'MT-MY Demeter (Feedermax)', 'DE-BE Siren (New Panamax)',
       'GB-IN Argo (U

In [4]:
container_dict = {}

for ship_inf in df.columns:
    for match in re.finditer(r'^[A-Za-z]{2}-[A-Za-z]{2}\ [-\u0041-\u005A\u0061-\u007A\u00AA\u00B5\u00BA\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u02C1\u02C6-\u02D1\u02E0-\u02E4\u02EC\u02EE\u0370-\u0374\u0376\u0377\u037A-\u037D\u0386\u0388-\u038A\u038C\u038E-\u03A1\u03A3-\u03F5\u03F7-\u0481\u048A-\u0527\u0531-\u0556\u0559\u0561-\u0587\u05D0-\u05EA\u05F0-\u05F2\u0620-\u064A\u066E\u066F\u0671-\u06D3\u06D5\u06E5\u06E6\u06EE\u06EF\u06FA-\u06FC\u06FF\u0710\u0712-\u072F\u074D-\u07A5\u07B1\u07CA-\u07EA\u07F4\u07F5\u07FA\u0800-\u0815\u081A\u0824\u0828\u0840-\u0858\u08A0\u08A2-\u08AC\u0904-\u0939\u093D\u0950\u0958-\u0961\u0971-\u0977\u0979-\u097F\u0985-\u098C\u098F\u0990\u0993-\u09A8\u09AA-\u09B0\u09B2\u09B6-\u09B9\u09BD\u09CE\u09DC\u09DD\u09DF-\u09E1\u09F0\u09F1\u0A05-\u0A0A\u0A0F\u0A10\u0A13-\u0A28\u0A2A-\u0A30\u0A32\u0A33\u0A35\u0A36\u0A38\u0A39\u0A59-\u0A5C\u0A5E\u0A72-\u0A74\u0A85-\u0A8D\u0A8F-\u0A91\u0A93-\u0AA8\u0AAA-\u0AB0\u0AB2\u0AB3\u0AB5-\u0AB9\u0ABD\u0AD0\u0AE0\u0AE1\u0B05-\u0B0C\u0B0F\u0B10\u0B13-\u0B28\u0B2A-\u0B30\u0B32\u0B33\u0B35-\u0B39\u0B3D\u0B5C\u0B5D\u0B5F-\u0B61\u0B71\u0B83\u0B85-\u0B8A\u0B8E-\u0B90\u0B92-\u0B95\u0B99\u0B9A\u0B9C\u0B9E\u0B9F\u0BA3\u0BA4\u0BA8-\u0BAA\u0BAE-\u0BB9\u0BD0\u0C05-\u0C0C\u0C0E-\u0C10\u0C12-\u0C28\u0C2A-\u0C33\u0C35-\u0C39\u0C3D\u0C58\u0C59\u0C60\u0C61\u0C85-\u0C8C\u0C8E-\u0C90\u0C92-\u0CA8\u0CAA-\u0CB3\u0CB5-\u0CB9\u0CBD\u0CDE\u0CE0\u0CE1\u0CF1\u0CF2\u0D05-\u0D0C\u0D0E-\u0D10\u0D12-\u0D3A\u0D3D\u0D4E\u0D60\u0D61\u0D7A-\u0D7F\u0D85-\u0D96\u0D9A-\u0DB1\u0DB3-\u0DBB\u0DBD\u0DC0-\u0DC6\u0E01-\u0E30\u0E32\u0E33\u0E40-\u0E46\u0E81\u0E82\u0E84\u0E87\u0E88\u0E8A\u0E8D\u0E94-\u0E97\u0E99-\u0E9F\u0EA1-\u0EA3\u0EA5\u0EA7\u0EAA\u0EAB\u0EAD-\u0EB0\u0EB2\u0EB3\u0EBD\u0EC0-\u0EC4\u0EC6\u0EDC-\u0EDF\u0F00\u0F40-\u0F47\u0F49-\u0F6C\u0F88-\u0F8C\u1000-\u102A\u103F\u1050-\u1055\u105A-\u105D\u1061\u1065\u1066\u106E-\u1070\u1075-\u1081\u108E\u10A0-\u10C5\u10C7\u10CD\u10D0-\u10FA\u10FC-\u1248\u124A-\u124D\u1250-\u1256\u1258\u125A-\u125D\u1260-\u1288\u128A-\u128D\u1290-\u12B0\u12B2-\u12B5\u12B8-\u12BE\u12C0\u12C2-\u12C5\u12C8-\u12D6\u12D8-\u1310\u1312-\u1315\u1318-\u135A\u1380-\u138F\u13A0-\u13F4\u1401-\u166C\u166F-\u167F\u1681-\u169A\u16A0-\u16EA\u1700-\u170C\u170E-\u1711\u1720-\u1731\u1740-\u1751\u1760-\u176C\u176E-\u1770\u1780-\u17B3\u17D7\u17DC\u1820-\u1877\u1880-\u18A8\u18AA\u18B0-\u18F5\u1900-\u191C\u1950-\u196D\u1970-\u1974\u1980-\u19AB\u19C1-\u19C7\u1A00-\u1A16\u1A20-\u1A54\u1AA7\u1B05-\u1B33\u1B45-\u1B4B\u1B83-\u1BA0\u1BAE\u1BAF\u1BBA-\u1BE5\u1C00-\u1C23\u1C4D-\u1C4F\u1C5A-\u1C7D\u1CE9-\u1CEC\u1CEE-\u1CF1\u1CF5\u1CF6\u1D00-\u1DBF\u1E00-\u1F15\u1F18-\u1F1D\u1F20-\u1F45\u1F48-\u1F4D\u1F50-\u1F57\u1F59\u1F5B\u1F5D\u1F5F-\u1F7D\u1F80-\u1FB4\u1FB6-\u1FBC\u1FBE\u1FC2-\u1FC4\u1FC6-\u1FCC\u1FD0-\u1FD3\u1FD6-\u1FDB\u1FE0-\u1FEC\u1FF2-\u1FF4\u1FF6-\u1FFC\u2071\u207F\u2090-\u209C\u2102\u2107\u210A-\u2113\u2115\u2119-\u211D\u2124\u2126\u2128\u212A-\u212D\u212F-\u2139\u213C-\u213F\u2145-\u2149\u214E\u2183\u2184\u2C00-\u2C2E\u2C30-\u2C5E\u2C60-\u2CE4\u2CEB-\u2CEE\u2CF2\u2CF3\u2D00-\u2D25\u2D27\u2D2D\u2D30-\u2D67\u2D6F\u2D80-\u2D96\u2DA0-\u2DA6\u2DA8-\u2DAE\u2DB0-\u2DB6\u2DB8-\u2DBE\u2DC0-\u2DC6\u2DC8-\u2DCE\u2DD0-\u2DD6\u2DD8-\u2DDE\u2E2F\u3005\u3006\u3031-\u3035\u303B\u303C\u3041-\u3096\u309D-\u309F\u30A1-\u30FA\u30FC-\u30FF\u3105-\u312D\u3131-\u318E\u31A0-\u31BA\u31F0-\u31FF\u3400-\u4DB5\u4E00-\u9FCC\uA000-\uA48C\uA4D0-\uA4FD\uA500-\uA60C\uA610-\uA61F\uA62A\uA62B\uA640-\uA66E\uA67F-\uA697\uA6A0-\uA6E5\uA717-\uA71F\uA722-\uA788\uA78B-\uA78E\uA790-\uA793\uA7A0-\uA7AA\uA7F8-\uA801\uA803-\uA805\uA807-\uA80A\uA80C-\uA822\uA840-\uA873\uA882-\uA8B3\uA8F2-\uA8F7\uA8FB\uA90A-\uA925\uA930-\uA946\uA960-\uA97C\uA984-\uA9B2\uA9CF\uAA00-\uAA28\uAA40-\uAA42\uAA44-\uAA4B\uAA60-\uAA76\uAA7A\uAA80-\uAAAF\uAAB1\uAAB5\uAAB6\uAAB9-\uAABD\uAAC0\uAAC2\uAADB-\uAADD\uAAE0-\uAAEA\uAAF2-\uAAF4\uAB01-\uAB06\uAB09-\uAB0E\uAB11-\uAB16\uAB20-\uAB26\uAB28-\uAB2E\uABC0-\uABE2\uAC00-\uD7A3\uD7B0-\uD7C6\uD7CB-\uD7FB\uF900-\uFA6D\uFA70-\uFAD9\uFB00-\uFB06\uFB13-\uFB17\uFB1D\uFB1F-\uFB28\uFB2A-\uFB36\uFB38-\uFB3C\uFB3E\uFB40\uFB41\uFB43\uFB44\uFB46-\uFBB1\uFBD3-\uFD3D\uFD50-\uFD8F\uFD92-\uFDC7\uFDF0-\uFDFB\uFE70-\uFE74\uFE76-\uFEFC\uFF21-\uFF3A\uFF41-\uFF5A\uFF66-\uFFBE\uFFC2-\uFFC7\uFFCA-\uFFCF\uFFD2-\uFFD7\uFFDA-\uFFDC]+ \(', ship_inf):
        company_name = ship_inf[match.span()[1]:-1]
        
    if not company_name in container_dict:
        container_dict[company_name] = (0, 0)
        
    container_dict[company_name] = (container_dict[company_name][0]+1, container_dict[company_name][1] + df[ship_inf].notnull().sum())

max_value = (0, 0)
for key, value in container_dict.items():
    actual_iter_max = value[1]/value[0]
    if actual_iter_max > max_value[1]:
        max_value = (key, actual_iter_max)
    
print(max_value)

('ULCV', 19668.090909090908)


Najwięcej kontenerów przewożą statki klasy ULCV

## PYTANIE NR. 3 Jaka jest średnia waga kontenera z przetworami owocowymi (A3) z dokładnością do 1 kg np: 1234? (zaokrąglane w górę) – 20 pkt

In [5]:
weight_objects=0
weight_sum=0

for ship_info in df.columns:
    weights = df[ship_info].str.extractall(r'[A-Za-z]{2}-[A-Za-z]{2}-\d+/(.+)/A3')
    weight_objects += weights.shape[0]    
    weights.reset_index(drop=True, inplace=True)
    weights = weights.astype(int)
    weight_sum += weights.sum(axis=0)

avg_weight = weight_sum/weight_objects

import math

math.ceil(avg_weight)

2924

Średnia waga kontenera z przetworami owocowymi (A3) TO 2924

## PYTANIE NR. 4 Która Polska firma globalnie przewozi najwięcej kontenerów? (nazwa firmy) – 25 pkt

In [6]:
polish_companys_dict = {}

for ship_info in df.columns:
    polish_companys = df[ship_info].str.extractall(r'[A-Za-z]{2}-[A-Za-z]{2}-\d+/\d{4}/[A-Z0-9]{2}@(.+)\.pl')
    polish_companys.reset_index(drop=True, inplace=True)
    for index, company in polish_companys.iterrows():
        if not company[0] in polish_companys_dict:
            polish_companys_dict[company[0]] = 0 
            
        polish_companys_dict[company[0]] += 1
        
        
best_polish_company = (0, 0)
for key, value in polish_companys_dict.items():
    if best_polish_company[1] < value:
        best_polish_company = (key, value)
        
        
print(best_polish_company)

('qdiku', 274)


Polska firma, która globalnie przewozi najwięcej kontenerów to qdiku

## PYTANIE NR. 5 Jakiego typu ładunek o największej wartości eksportują Niemieckie firmy z Niemiec? (wartość to stosunek ceny do wagi) – 25 pkt

In [7]:
max_value = (0, 0)
for ship_info in df.columns:
    data = df[ship_info].str.extractall(r'DE-[A-Za-z]{2}-\d+/(\d{4})/([A-Z0-9]{2})@.+\.de/(\d+)')
    if not data.empty:
        data.reset_index(drop=True, inplace=True)
        data[[0,2]] = data[[0,2]].astype(int)
        data['value'] = data[2]/data[0]
        idmax = data['value'].idxmax()
        actual_iter_max_value = (data.iloc[idmax,1], data.iloc[idmax, 3])
        if max_value[1] < actual_iter_max_value[1]:
            max_value = actual_iter_max_value
        elif max_value[1] == actual_iter_max_value[1]:
            print('THERE ARE 2 SAME VALUES! WARNING')
        
print(max_value)

('C1', 83.74871355060034)


Ładunek o największej wartości, eksportowany z Niemiec przez niemiecką firme jest typu C1